File: core/recorder.py
import time
import threading
import os
from typing import List, Callable, Optional, Dict, Any
from dataclasses import dataclass
from enum import Enum
from pynput import mouse, keyboard
from .actions import Action, ActionType


class RecordState(Enum):
    IDLE = "idle"
    RECORDING = "recording"
    PAUSED = "paused"


@dataclass
class RecordConfig:
    record_mouse_move: bool = False
    record_mouse_click: bool = True
    record_mouse_scroll: bool = True
    record_keyboard: bool = True
    min_move_distance: int = 10
    move_sample_interval: float = 0.1
    ignore_last_click: bool = True
    record_as_image_click: bool = False
    image_capture_size: int = 30


class Recorder:
    def __init__(self, config: Optional[RecordConfig] = None):
        self.config = config or RecordConfig()
        self.state = RecordState.IDLE
        self.actions: List[Action] = []
        self.last_mouse_pos: Optional[tuple] = None
        self.last_move_time: float = 0
        self.start_time: float = 0
        self.last_action_time: float = 0
        self._stop_time: float = 0
        self._stopping: bool = False
        self._lock = threading.Lock()
        
        self._mouse_listener: Optional[mouse.Listener] = None
        self._keyboard_listener: Optional[keyboard.Listener] = None
        
        self._callbacks: Dict[str, List[Callable]] = {
            'on_action_recorded': [],
            'on_state_changed': [],
            'on_error': [],
            'on_last_action_removed': []
        }
        
        self._recorded_keys: List[str] = []
        self._is_recording_text: bool = False
        self._text_start_time: float = 0
    
    def add_callback(self, event: str, callback: Callable):
        if event in self._callbacks:
            self._callbacks[event].append(callback)
    
    def remove_callback(self, event: str, callback: Callable):
        if event in self._callbacks and callback in self._callbacks[event]:
            self._callbacks[event].remove(callback)
    
    def _emit(self, event: str, *args, **kwargs):
        for callback in self._callbacks.get(event, []):
            try:
                callback(*args, **kwargs)
            except Exception as e:
                print(f"Callback error: {e}")
    
    def start(self):
        if self.state == RecordState.RECORDING:
            return
        
        self.actions = []
        self.state = RecordState.RECORDING
        self.start_time = time.time()
        self.last_action_time = self.start_time
        self.last_mouse_pos = None
        self._recorded_keys = []
        self._is_recording_text = False
        
        self._mouse_listener = mouse.Listener(
            on_move=self._on_mouse_move,
            on_click=self._on_mouse_click,
            on_scroll=self._on_mouse_scroll
        )
        
        self._keyboard_listener = keyboard.Listener(
            on_press=self._on_key_press,
            on_release=self._on_key_release
        )
        
        self._mouse_listener.start()
        self._keyboard_listener.start()
        
        self._emit('on_state_changed', self.state)
    
    def stop(self):
        with self._lock:
            if self.state == RecordState.IDLE:
                return
            
            self._stopping = True
            self.state = RecordState.IDLE
        
        if self._mouse_listener:
            self._mouse_listener.stop()
            self._mouse_listener = None
        
        if self._keyboard_listener:
            self._keyboard_listener.stop()
            self._keyboard_listener = None
        
        removed_last = False
        with self._lock:
            if self._is_recording_text and self._recorded_keys:
                self._flush_text_input()
            
            if self.actions and self.config.ignore_last_click:
                last_action = self.actions[-1]
                if last_action.action_type in (ActionType.MOUSE_CLICK, ActionType.IMAGE_CLICK):
                    self.actions.pop()
                    removed_last = True
            
            self._stopping = False
        
        if removed_last:
            self._emit('on_last_action_removed')
        
        self._emit('on_state_changed', self.state)
    
    def pause(self):
        if self.state == RecordState.RECORDING:
            self.state = RecordState.PAUSED
            self._emit('on_state_changed', self.state)
    
    def resume(self):
        if self.state == RecordState.PAUSED:
            self.state = RecordState.RECORDING
            self._emit('on_state_changed', self.state)
    
    def _get_elapsed_time(self) -> float:
        return time.time() - self.last_action_time
    
    def _add_action(self, action: Action):
        with self._lock:
            if self.state != RecordState.RECORDING:
                return
            
            elapsed = self._get_elapsed_time()
            action.delay_before = elapsed
            action.description = action._generate_description()
            self.actions.append(action)
            self.last_action_time = time.time()
        self._emit('on_action_recorded', action)
    
    def _on_mouse_move(self, x: int, y: int):
        if self.state != RecordState.RECORDING:
            return
        
        if not self.config.record_mouse_move:
            return
        
        current_time = time.time()
        
        if current_time - self.last_move_time < self.config.move_sample_interval:
            return
        
        if self.last_mouse_pos:
            distance = ((x - self.last_mouse_pos[0]) ** 2 + (y - self.last_mouse_pos[1]) ** 2) ** 0.5
            if distance < self.config.min_move_distance:
                return
        
        self.last_mouse_pos = (x, y)
        self.last_move_time = current_time
        
        action = Action(
            action_type=ActionType.MOUSE_MOVE,
            params={'x': x, 'y': y, 'duration': 0.0}
        )
        self._add_action(action)
    
    def _on_mouse_click(self, x: int, y: int, button: mouse.Button, pressed: bool):
        if not pressed:
            return
        
        if not self.config.record_mouse_click:
            return
        
        with self._lock:
            if self.state != RecordState.RECORDING:
                return
            
            if self._stopping:
                return
            
            if self._is_recording_text:
                self._flush_text_input()
            
            button_name = 'left' if button == mouse.Button.left else 'right' if button == mouse.Button.right else 'middle'
            
            if self.config.record_as_image_click:
                image_path = self._capture_click_region(x, y)
                if image_path:
                    action = Action(
                        action_type=ActionType.IMAGE_CLICK,
                        params={'image_path': image_path, 'confidence': 0.9}
                    )
                else:
                    action = Action(
                        action_type=ActionType.MOUSE_CLICK,
                        params={'x': x, 'y': y, 'button': button_name, 'clicks': 1}
                    )
            else:
                action = Action(
                    action_type=ActionType.MOUSE_CLICK,
                    params={'x': x, 'y': y, 'button': button_name, 'clicks': 1}
                )
            
            elapsed = self._get_elapsed_time()
            action.delay_before = elapsed
            action.description = action._generate_description()
            self.actions.append(action)
            self.last_action_time = time.time()
        
        self._emit('on_action_recorded', action)
    
    def _capture_click_region(self, x: int, y: int) -> Optional[str]:
        try:
            import pyautogui
            
            size = self.config.image_capture_size
            region = (
                max(0, x - size // 2),
                max(0, y - size // 2),
                size,
                size
            )
            
            screenshot = pyautogui.screenshot(region=region)
            
            images_dir = os.path.join(os.path.expanduser('~'), '.simpleRPA', 'images')
            os.makedirs(images_dir, exist_ok=True)
            
            filename = f"click_{int(time.time() * 1000)}.png"
            filepath = os.path.join(images_dir, filename)
            
            screenshot.save(filepath)
            
            return filepath
        except Exception as e:
            print(f"截图失败: {e}")
            return None
    
    def _on_mouse_scroll(self, x: int, y: int, dx: int, dy: int):
        if self.state != RecordState.RECORDING:
            return
        
        if not self.config.record_mouse_scroll:
            return
        
        if self._is_recording_text:
            self._flush_text_input()
        
        action = Action(
            action_type=ActionType.MOUSE_SCROLL,
            params={'clicks': dy, 'x': x, 'y': y}
        )
        self._add_action(action)
    
    def _on_key_press(self, key):
        if self.state != RecordState.RECORDING:
            return
        
        if not self.config.record_keyboard:
            return
        
        try:
            if hasattr(key, 'char') and key.char:
                if not self._is_recording_text:
                    self._is_recording_text = True
                    self._text_start_time = time.time()
                self._recorded_keys.append(key.char)
            else:
                if self._is_recording_text:
                    self._flush_text_input()
                
                key_name = self._get_key_name(key)
                if key_name:
                    modifier_keys = ['ctrl', 'alt', 'shift', 'cmd', 'win']
                    if key_name.lower() in modifier_keys:
                        return
                    
                    action = Action(
                        action_type=ActionType.KEY_PRESS,
                        params={'key': key_name}
                    )
                    self._add_action(action)
        except Exception as e:
            print(f"Key press error: {e}")
    
    def _on_key_release(self, key):
        pass
    
    def _get_key_name(self, key) -> Optional[str]:
        key_map = {
            keyboard.Key.space: 'space',
            keyboard.Key.enter: 'enter',
            keyboard.Key.tab: 'tab',
            keyboard.Key.backspace: 'backspace',
            keyboard.Key.delete: 'delete',
            keyboard.Key.esc: 'escape',
            keyboard.Key.up: 'up',
            keyboard.Key.down: 'down',
            keyboard.Key.left: 'left',
            keyboard.Key.right: 'right',
            keyboard.Key.home: 'home',
            keyboard.Key.end: 'end',
            keyboard.Key.page_up: 'pageup',
            keyboard.Key.page_down: 'pagedown',
            keyboard.Key.caps_lock: 'capslock',
            keyboard.Key.f1: 'f1',
            keyboard.Key.f2: 'f2',
            keyboard.Key.f3: 'f3',
            keyboard.Key.f4: 'f4',
            keyboard.Key.f5: 'f5',
            keyboard.Key.f6: 'f6',
            keyboard.Key.f7: 'f7',
            keyboard.Key.f8: 'f8',
            keyboard.Key.f9: 'f9',
            keyboard.Key.f10: 'f10',
            keyboard.Key.f11: 'f11',
            keyboard.Key.f12: 'f12',
        }
        
        if key in key_map:
            return key_map[key]
        
        if hasattr(key, 'name'):
            return key.name
        
        return None
    
    def _flush_text_input(self):
        if not self._recorded_keys:
            self._is_recording_text = False
            return
        
        text = ''.join(self._recorded_keys)
        action = Action(
            action_type=ActionType.KEY_TYPE,
            params={'text': text, 'interval': 0.0}
        )
        self._add_action(action)
        
        self._recorded_keys = []
        self._is_recording_text = False
    
    def get_actions(self) -> List[Action]:
        if self._is_recording_text and self._recorded_keys:
            self._flush_text_input()
        return self.actions.copy()
    
    def clear_actions(self):
        self.actions = []
        self.last_action_time = time.time()
    
    def is_recording(self) -> bool:
        return self.state == RecordState.RECORDING
    
    def get_state(self) -> RecordState:
        return self.state

File: rust-workspace/crates/simpleRPA-core/src/recorder.rs
use crate::actions::{Action, ActionType};
use std::sync::{Arc, Mutex};
use std::time::Instant;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum RecordState {
    Idle,
    Recording,
    Paused,
}

pub struct RecordConfig {
    pub record_mouse_move: bool,
    pub record_mouse_click: bool,
    pub record_mouse_scroll: bool,
    pub record_keyboard: bool,
    pub min_move_distance: i32,
    pub move_sample_interval: f64,
    pub ignore_last_click: bool,
}

impl Default for RecordConfig {
    fn default() -> Self {
        Self {
            record_mouse_move: false,
            record_mouse_click: true,
            record_mouse_scroll: true,
            record_keyboard: true,
            min_move_distance: 10,
            move_sample_interval: 0.1,
            ignore_last_click: true,
        }
    }
}

pub struct Recorder {
    config: RecordConfig,
    state: Arc<Mutex<RecordState>>,
    actions: Arc<Mutex<Vec<Action>>>,
    last_mouse_pos: Arc<Mutex<Option<(i32, i32)>>>,
    last_move_time: Arc<Mutex<Instant>>,
    start_time: Arc<Mutex<Instant>>,
    last_action_time: Arc<Mutex<Instant>>,
    recorded_keys: Arc<Mutex<Vec<char>>>,
    is_recording_text: Arc<Mutex<bool>>,

    on_action_recorded: Option<Box<dyn Fn(&Action) + Send>>,
    on_state_changed: Option<Box<dyn Fn(RecordState) + Send>>,
}

impl Recorder {
    pub fn new(config: Option<RecordConfig>) -> Self {
        let now = Instant::now();
        Self {
            config: config.unwrap_or_default(),
            state: Arc::new(Mutex::new(RecordState::Idle)),
            actions: Arc::new(Mutex::new(Vec::new())),
            last_mouse_pos: Arc::new(Mutex::new(None)),
            last_move_time: Arc::new(Mutex::new(now)),
            start_time: Arc::new(Mutex::new(now)),
            last_action_time: Arc::new(Mutex::new(now)),
            recorded_keys: Arc::new(Mutex::new(Vec::new())),
            is_recording_text: Arc::new(Mutex::new(false)),
            on_action_recorded: None,
            on_state_changed: None,
        }
    }

    pub fn set_on_action_recorded<F: Fn(&Action) + Send + 'static>(&mut self, f: F) {
        self.on_action_recorded = Some(Box::new(f));
    }

    pub fn set_on_state_changed<F: Fn(RecordState) + Send + 'static>(&mut self, f: F) {
        self.on_state_changed = Some(Box::new(f));
    }

    pub fn set_config(&mut self, config: RecordConfig) {
        self.config = config;
    }

    pub fn start(&self) {
        let mut state = self.state.lock().unwrap();
        if *state == RecordState::Recording {
            return;
        }

        self.actions.lock().unwrap().clear();
        *state = RecordState::Recording;
        *self.start_time.lock().unwrap() = Instant::now();
        *self.last_action_time.lock().unwrap() = Instant::now();
        *self.last_mouse_pos.lock().unwrap() = None;
        self.recorded_keys.lock().unwrap().clear();
        *self.is_recording_text.lock().unwrap() = false;

        if let Some(ref f) = self.on_state_changed {
            f(RecordState::Recording);
        }
    }

    pub fn stop(&self) -> Vec<Action> {
        let mut state = self.state.lock().unwrap();
        if *state == RecordState::Idle {
            return Vec::new();
        }

        *state = RecordState::Idle;

        if *self.is_recording_text.lock().unwrap() {
            self.flush_text_input();
        }

        let mut actions = self.actions.lock().unwrap();
        if self.config.ignore_last_click {
            if let Some(last) = actions.last() {
                if matches!(
                    last.action_type,
                    ActionType::MouseClick | ActionType::ImageClick
                ) {
                    actions.pop();
                }
            }
        }

        if let Some(ref f) = self.on_state_changed {
            f(RecordState::Idle);
        }

        actions.clone()
    }

    pub fn pause(&self) {
        let mut state = self.state.lock().unwrap();
        if *state == RecordState::Recording {
            *state = RecordState::Paused;
            if let Some(ref f) = self.on_state_changed {
                f(RecordState::Paused);
            }
        }
    }

    pub fn resume(&self) {
        let mut state = self.state.lock().unwrap();
        if *state == RecordState::Paused {
            *state = RecordState::Recording;
            if let Some(ref f) = self.on_state_changed {
                f(RecordState::Recording);
            }
        }
    }

    pub fn get_actions(&self) -> Vec<Action> {
        self.actions.lock().unwrap().clone()
    }

    pub fn clear_actions(&self) {
        self.actions.lock().unwrap().clear();
        *self.last_action_time.lock().unwrap() = Instant::now();
    }

    pub fn is_recording(&self) -> bool {
        *self.state.lock().unwrap() == RecordState::Recording
    }

    pub fn on_mouse_move(&self, x: i32, y: i32) {
        if *self.state.lock().unwrap() != RecordState::Recording {
            return;
        }
        if !self.config.record_mouse_move {
            return;
        }

        let now = Instant::now();
        let last_move = *self.last_move_time.lock().unwrap();
        if now.duration_since(last_move).as_secs_f64() < self.config.move_sample_interval {
            return;
        }

        let mut last_pos = self.last_mouse_pos.lock().unwrap();
        if let Some(pos) = *last_pos {
            let distance = ((x - pos.0).pow(2) + (y - pos.1).pow(2)) as f64;
            if (distance as i32) < self.config.min_move_distance {
                return;
            }
        }
        *last_pos = Some((x, y));
        *self.last_move_time.lock().unwrap() = now;

        let mut action = Action::new(ActionType::MouseMove);
        action.params.insert("x".into(), serde_json::json!(x));
        action.params.insert("y".into(), serde_json::json!(y));
        action
            .params
            .insert("duration".into(), serde_json::json!(0.0));
        self.add_action(action);
    }

    pub fn on_mouse_click(&self, x: i32, y: i32, button: &str) {
        if *self.state.lock().unwrap() != RecordState::Recording {
            return;
        }
        if !self.config.record_mouse_click {
            return;
        }

        if *self.is_recording_text.lock().unwrap() {
            self.flush_text_input();
        }

        let elapsed = self.get_elapsed_time();
        let mut action = Action::new(ActionType::MouseClick);
        action.params.insert("x".into(), serde_json::json!(x));
        action.params.insert("y".into(), serde_json::json!(y));
        action
            .params
            .insert("button".into(), serde_json::json!(button));
        action.params.insert("clicks".into(), serde_json::json!(1));
        action.delay_before = elapsed;
        action.description = action.generate_description();
        self.actions.lock().unwrap().push(action.clone());
        *self.last_action_time.lock().unwrap() = Instant::now();

        if let Some(ref f) = self.on_action_recorded {
            f(&action);
        }
    }

    pub fn on_mouse_scroll(&self, x: i32, y: i32, _dx: i32, dy: i32) {
        if *self.state.lock().unwrap() != RecordState::Recording {
            return;
        }
        if !self.config.record_mouse_scroll {
            return;
        }

        if *self.is_recording_text.lock().unwrap() {
            self.flush_text_input();
        }

        let mut action = Action::new(ActionType::MouseScroll);
        action.params.insert("clicks".into(), serde_json::json!(dy));
        action.params.insert("x".into(), serde_json::json!(x));
        action.params.insert("y".into(), serde_json::json!(y));
        self.add_action(action);
    }

    pub fn on_key_press(&self, key: char) {
        if *self.state.lock().unwrap() != RecordState::Recording {
            return;
        }
        if !self.config.record_keyboard {
            return;
        }

        if !*self.is_recording_text.lock().unwrap() {
            *self.is_recording_text.lock().unwrap() = true;
        }
        self.recorded_keys.lock().unwrap().push(key);
    }

    fn flush_text_input(&self) {
        let keys: Vec<char> = {
            let mut recorded = self.recorded_keys.lock().unwrap();
            if recorded.is_empty() {
                *self.is_recording_text.lock().unwrap() = false;
                return;
            }
            recorded.drain(..).collect()
        };

        let text: String = keys.iter().collect();
        let mut action = Action::new(ActionType::KeyType);
        action.params.insert("text".into(), serde_json::json!(text));
        action
            .params
            .insert("interval".into(), serde_json::json!(0.0));
        self.add_action(action);

        *self.is_recording_text.lock().unwrap() = false;
    }

    fn add_action(&self, mut action: Action) {
        let state = self.state.lock().unwrap();
        if *state != RecordState::Recording {
            return;
        }

        let elapsed = self.get_elapsed_time();
        action.delay_before = elapsed;
        action.description = action.generate_description();
        self.actions.lock().unwrap().push(action.clone());
        *self.last_action_time.lock().unwrap() = Instant::now();

        drop(state);
        if let Some(ref f) = self.on_action_recorded {
            f(&action);
        }
    }

    fn get_elapsed_time(&self) -> f64 {
        self.last_action_time
            .lock()
            .unwrap()
            .elapsed()
            .as_secs_f64()
    }
}
